osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs

```diff
@@ -20,51 +20,165 @@ public class Speed : OsuStrainSkill
         private const double pi_over_4 = Math.PI / 4;
         private const double pi_over_2 = Math.PI / 2;
 
-        protected override double SkillMultiplier => 1400;
-        protected override double StrainDecayBase => 0.3;
+        private const double rhythm_multiplier = 2.0;
+        private const int history_time_max = 3000; // 3 seconds of calculatingRhythmBonus max.
+
+        private double skillMultiplier => 1375;
+        private double strainDecayBase => 0.3;
+
+        private double currentTapStrain = 1;
+        private double currentMovementStrain = 1;
+        private double currentRhythm = 1;
+
         protected override int ReducedSectionCount => 5;
         protected override double DifficultyMultiplier => 1.04;
 
         private const double min_speed_bonus = 75; // ~200BPM
         private const double max_speed_bonus = 45; // ~330BPM
         private const double speed_balancing_factor = 40;
 
+        protected override int HistoryLength => 32;
+
         public Speed(Mod[] mods)
             : base(mods)
         {
         }
 
-        protected override double StrainValueOf(DifficultyHitObject current)
+        private bool isRatioEqual(double ratio, double a, double b)
+        {
+            return a + 15 > ratio * b && a - 15 < ratio * b;
+        }
```

Comment:
Seems like this could be replaced with `Precision.AlmostEquals(a, b, 15)` or similar.

Reply:
Done.